File: Solutions/33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.dart
class Solution {
  int search(List<int> nums, int target) {
    // Initialize two pointers for binary search
    int left = 0;                // Start of the array
    int right = nums.length - 1; // End of the array

    // Binary search loop
    while (left <= right) {
      // Find the middle index
      int mid = left + ((right - left) ~/ 2);

      // Check if the middle element is the target
      if (nums[mid] == target) {
        return mid; // Target found, return index
      }

      // Determine which side is sorted
      if (nums[left] <= nums[mid]) {
        // Left side from 'left' to 'mid' is sorted
        if (nums[left] <= target && target < nums[mid]) {
          // Target is within the sorted left half
          right = mid - 1; // Narrow search to left half
        } else {
          // Target is not in left half
          left = mid + 1; // Search in the right half
        }
      } else {
        // Right side from 'mid' to 'right' is sorted
        if (nums[mid] < target && target <= nums[right]) {
          // Target is within the sorted right half
          left = mid + 1; // Narrow search to right half
        } else {
          // Target is not in right half
          right = mid - 1; // Search in the left half
        }
      }
    }

    // Target was not found in the array
    return -1;
  }
}

File: Solutions/33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.ts
function search(nums: number[], target: number): number {
    // We use binary search because:
    // - The array is sorted
    // - But it may be rotated
    // - We need O(log n) time complexity

    // Left pointer
    let left = 0;

    // Right pointer
    let right = nums.length - 1;

    // Continue searching while the window is valid
    while (left <= right) {

        // Find middle index
        // Math.floor prevents decimals
        const mid = Math.floor((left + right) / 2);

        // If we found the target, return its index
        if (nums[mid] === target) {
            return mid;
        }

        // IMPORTANT IDEA:
        // At least ONE HALF is always sorted

        // Check if LEFT HALF is sorted
        if (nums[left] <= nums[mid]) {

            // Now check if target lies inside the sorted left half
            if (target >= nums[left] && target < nums[mid]) {

                // Target is inside left half
                // Move right pointer inward
                right = mid - 1;

            } else {

                // Target is NOT in left half
                // Search the right half
                left = mid + 1;
            }

        } else {

            // Otherwise RIGHT HALF must be sorted

            // Check if target lies inside sorted right half
            if (target > nums[mid] && target <= nums[right]) {

                // Target is inside right half
                // Ignore left side
                left = mid + 1;

            } else {

                // Target is NOT in right half
                // Search left half
                right = mid - 1;
            }
        }
    }

    // If loop finishes, target does not exist
    return -1;
}